### PR TITLE
rbd-mirror: create and export replication perf counters to mgr 

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/21392
Signed-off-by: Mykola Golub <mgolub@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by trociny** [comment]
An example of prometheus module output for "mirror/test1" image:
```
ceph_rbd_mirror_mirror/test1_replay{ceph_daemon="rbd-mirror.4534"} 18.0
ceph_rbd_mirror_mirror/test1_replay_bytes{ceph_daemon="rbd-mirror.4534"} 66288.0
ceph_rbd_mirror_mirror/test1_replay_latency_sum{ceph_daemon="rbd-mirror.4534"} 0.028574119
ceph_rbd_mirror_mirror/test1_replay_latency_count{ceph_daemon="rbd-mirror.4534"} 18.0
```

**Comment by trociny** [comment]
Added a commit for prometheus module to export rbd-mirror metadata. It looks like this:
```
# HELP ceph_rbd_mirror_metadata RBD Mirror Metadata
# TYPE ceph_rbd_mirror_metadata untyped
ceph_rbd_mirror_metadata{ceph_daemon="rbd-mirror.4534",id="mirror.0",instance_id="4534",hostname="adonis",ceph_version="ceph version 14.0.1-2338-gc2623ad872 (c2623ad87223e10b7d091763d05cd3dd438f1e76) nautilus (dev)"} 1.0
```

File: src/common/options.cc [hunk 1]
```diff
@@ -7089,6 +7089,15 @@ static std::vector<Option> get_rbd_mirror_options() {
     Option("rbd_mirror_image_policy_rebalance_timeout", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
     .set_default(0)
     .set_description("number of seconds policy should be idle before trigerring reshuffle (rebalance) of images"),
+
+    Option("rbd_mirror_perf_stats_prio", Option::TYPE_INT, Option::LEVEL_ADVANCED)
+    .set_default((int64_t)PerfCountersBuilder::PRIO_USEFUL)
+    .set_description("Priority level for mirror daemon replication perf counters")
+    .set_long_description("The daemon will send perf counter data to the manager"
```
Comment (dillaman) [inline src/common/options.cc]: Nit: missing space at end of line

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 2]
```diff
@@ -56,6 +56,15 @@ std::ostream &operator<<(std::ostream &os,
 
 namespace {
 
+// Performance counters
+enum {
+  l_rbd_mirror_first = 26000,
```
Comment (dillaman) [inline src/tools/rbd_mirror/ImageReplayer.cc]: Nit: need unique ID range -- perhaps 27000?

File: src/tools/rbd_mirror/ImageReplayer.h [hunk 3]
```diff
@@ -318,6 +319,7 @@ class ImageReplayer {
   bool m_manual_stop = false;
 
   AdminSocketHook *m_asok_hook = nullptr;
+  PerfCounters *m_perfcounters = nullptr;
```
Comment (dillaman) [inline src/tools/rbd_mirror/ImageReplayer.h]: Nit: ```m_perf_counters```

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 4]
```diff
@@ -56,6 +56,15 @@ std::ostream &operator<<(std::ostream &os,
 
 namespace {
 
+// Performance counters
```
Comment (dillaman) [inline src/tools/rbd_mirror/ImageReplayer.cc]: Nit: move to ```Types.h```

**Comment by trociny** [comment]
@dillaman 
> It might be nice to have a rolled up aggregate metric for all images within a given daemon (i.e. rbd_mirror_ perf counter path).

Is the last commit I added looks like what you wanted? Other comments have been addressed too.

**Pull request merged**